src/planning.py

```diff
@@ -287,19 +288,33 @@ def _run_low_level_search(
 
 def _update_nsrts_with_failure(
         discovered_failure: _DiscoveredFailure,
-        ground_nsrts: Collection[_GroundNSRT]) -> None:
+        ground_nsrts: List[_GroundNSRT]
+        ) -> List[_GroundNSRT]:
     """Update the given set of ground_nsrts based on the given
-    DiscoveredFailure.
+    DiscoveredFailure. Returns a new list of ground NSRTs to replace the input
+    one, where all ground NSRTs that need modification are replaced with new
+    ones (because _GroundNSRTs are frozen).
     """
+    new_ground_nsrts = []
     for obj in discovered_failure.env_failure.offending_objects:
         atom = GroundAtom(Predicate(_NOT_CAUSES_FAILURE, [obj.type],
                                     _classifier=lambda s, o: False), [obj])
-        # Update the preconditions of the failing NSRT.
-        discovered_failure.failing_nsrt.preconditions.add(atom)
-        # Update the effects of all nsrts that use this object.
-        for nsrt in ground_nsrts:
-            if obj in nsrt.objects:
-                nsrt.add_effects.add(atom)
+        for ground_nsrt in ground_nsrts:
+            # Update the preconditions of the failing NSRT.
+            if ground_nsrt == discovered_failure.failing_nsrt:
+                new_ground_nsrt = ground_nsrt.copy_with(
+                    preconditions=ground_nsrt.preconditions | {atom})
+            # Update the effects of all nsrts that use this object.
+            # Note that this is an elif rather than an if, because it would
```

Comment:
This is a slight logic change, but it makes sense to me. Verify that cluttered_table continues to work (`python src/main.py --env cluttered_table --approach oracle --seed 0` should give >45/50).

Reply:
yup I did